File: 13-roman-to-integer/13-roman-to-integer.scala
import scala.collection.mutable._

object Solution {
    def romanToInt(s: String): Int = {
        val map = Map({'I' -> 1}, {'V' -> 5}, {'X' -> 10}, {'L' -> 50}, {'C' -> 100}, {'D' -> 500}, {'M' -> 1000})
        var res = 0
        
        for(i <- 0 to s.length - 1) {
            val prev = if (i == 0) 0 else map(s.charAt(i - 1))
            val cur = map(s.charAt(i))
            
            if (prev * 5 == cur || prev * 10 == cur) {
                res -= 2 * prev
            }
            
            res += cur
        }
        
        res
    }
}

File: 13-roman-to-integer/13-roman-to-integer.java
class Solution {
    public int romanToInt(String s) {
        Map<Character, Integer> valueMap = Map.of(
            'I', 1,
            'V', 5,
            'X', 10,
            'L', 50,
            'C', 100,
            'D', 500,
            'M', 1000
        );
        
        int result = 0;
        
        for(int i = 0; i < s.length(); i++) {
            int v = valueMap.get(s.charAt(i));
            result += v;
            
            if (i > 0) {
                int pv = valueMap.get(s.charAt(i - 1));
                if (pv * 5 == v || pv * 10 == v) {
                    result -= 2 * pv;
                }
            }
        }
        
        return result;
    }
}